imxrt-hal/src/adc.rs

```diff
@@ -0,0 +1,275 @@
+//! ADC
+//!
+//! This ADC driver supports `embedded_hal`'s ADC traits
+//!
+//! # Example
+//! ```no_run
+//! use imxrt_hal::{self, adc};
+//! use embedded_hal::adc::OneShot;
+//!
+//! let mut peripherals = imxrt_hal::Peripherals::take().unwrap();
+//! let (adc1_builder, _) = peripherals.adc.clock(&mut peripherals.ccm.handle);
+//!
+//! let mut adc1 = adc1_builder.build(adc::ClockSelect::default(), adc::ClockDivision::default());
+//! let mut a1 = adc::AnalogInput::new(peripherals.iomuxc.ad_b1.p02, &adc1);
+//!
+//! let reading: u16 = adc1.read(&mut a1).unwrap();
+//!```
+
+use crate::ccm;
+use crate::iomuxc::adc::{prepare_adc_pin, Pin, ADC1, ADC2};
+use crate::iomuxc::{adc, consts::Unsigned};
+use crate::ral;
+use core::marker::PhantomData;
+use embedded_hal::adc::{Channel, OneShot};
+
+/// The clock input for an ADC
+#[allow(non_camel_case_types)]
+pub enum ClockSelect {
+    IPG,   // IPG clock
+    IPG_2, // IPG clock / 2
+    ADACK, // Asynchronous clock
+}
+
+/// How much to divide the clock input
+pub enum ClockDivision {
+    Div1, // Input clock / 1
+    Div2, // Input clock / 2
+    Div4, // Input clock / 4
+    Div8, // Input clock / 8
+}
+
+impl ClockSelect {
+    pub fn default() -> Self {
+        ClockSelect::ADACK
+    }
+}
+
+impl ClockDivision {
+    pub fn default() -> Self {
+        ClockDivision::Div2
+    }
+}
+
+/// Conversion speeds done by clock cycles
+pub enum ConversionSpeed {
+    Slow,     // 25 ADC clock cycles (24 on imxrt102x)
+    Medium,   // 17 ADC clock cycles (16 on imxrt102x)
+    Fast,     // 9 ADC clock cycles (8 on imxrt102x)
+    VeryFast, // 3 ADC clock cycles (2 on imxrt102x)
+}
+
+/// Denotes how much hardware averaging to do
+pub enum AveragingCount {
+    Avg1,
+    Avg4,
+    Avg8,
+    Avg16,
+    Avg32,
+}
+
+// Specifies the resolution the ADC
+pub enum ResolutionBits {
+    Res8,
+    Res10,
+    Res12,
+}
+
+/// A pin representing an analog input for a particular ADC
+pub struct AnalogInput<ADCx, P> {
+    _module: PhantomData<ADCx>,
+    _pin: P,
+}
+
+impl<P, ADCx> Channel<ADCx> for AnalogInput<ADCx, P>
+where
+    P: Pin<ADCx>,
+    ADCx: adc::ADC,
+{
+    type ID = u16;
+
+    fn channel() -> Self::ID {
+        <P as Pin<ADCx>>::Input::U16
+    }
+}
+
+impl<P, ADCx> AnalogInput<ADCx, P>
+where
+    P: Pin<ADCx>,
+    ADCx: adc::ADC,
+{
+    /// Creates a new analog input pin
+    pub fn new(mut pin: P, _adc: &ADC<ADCx>) -> Self {
+        prepare_adc_pin(&mut pin);
+        Self {
+            _module: PhantomData,
+            _pin: pin,
+        }
+    }
+}
```

Comment:
Should we add a `release` method to re-acquire the pad? Seems fitting with #20.
```suggestion
    }
    
    /// Release the ADC input, returning the underlying hardware pin
    pub fn release(self) -> P {
        self.pin
    }
}
```

Reply:
Sounds good to me. How should we "unprepare" the pin? An `unprepare` function in `imxrt-iomuxc`?

Reply:
Good thought! There hasn't been a need for an `unprepare` function yet. We could unprepare to the reset state, but that would require us to specify the reset states in code...

We can release the pad in an unspecified state. If a user wants to re-use the pad for something else, they will call a different `prepare` function on the pad.